### ISSUE TITLE
NVIDIA SHIELD TV simultaneous button presses not working

### ISSUE THREAD

**Issue by ReploidX** [libretro/RetroArch#8121]
## Description

This seems to be the same as this issue from 2016, but is on a current SHIELD TV and RetroArch.
https://github.com/libretro/RetroArch/issues/2043

### Expected behavior

Pressing two or three buttons should trigger a different move than one button. I've mainly been using fighting games to test this.

### Actual behavior

The majority of the time it just reads one button. Rarely will it read two or three.

Specifically SFZ2A in FBA, with moves like Zangief's lariat, or any super moves with two or three buttons.

Using Run-Ahead seems to help, but causes other problems like skipping frames or other visual glitches.

The SHIELD TV can detect the same buttons simultaneously when using GameStream or Steam Link. For example SFA2 in Street Fighter 30th Anniversary Collection works fine with the same controllers.

I tried these controllers:
- A wireless “gen 1” DualShock 4
- Fighting Commander for PlayStation 4 & 3 USB controller (mode settings didn’t seem to matter, but I usually leave it in PC and d-pad)
- SLS Sega Saturn USB controller

Here is a recent thread started by someone else on the Libretro forum who seems to have the same problem:
https://forums.libretro.com/t/nvidia-shield-tv-simultaneous-button-presses-bug/20095/2

### Steps to reproduce the bug

Play a game that uses simultaneous button presses. This seems to happen in MAME also - not just FBA.

### Version/Commit

- RetroArch: 1.7.5
- RetroArch64: 1.7.5

### Environment information

- OS: SHIELD EXPERIENCE 7.2

**Comment by orbea** [comment]
Can you reproduce this with the retropad builtin core? It works fine here.

**Comment by fr500** [comment]
I can't repro this:

https://streamable.com/oax9b

**Comment by ReploidX** [comment]
> I can't repro this:
> 
> https://streamable.com/oax9b

The menu I see looks the same as the XMB-style menu on desktop - not like the one at the end of the video. Is that video from a SATV? I haven't changed any interface settings.

I'll be able to post more details about version numbers probably later today, and will try SFA3 and some other games/cores.

On the other forum post someone said they have this [problem with the INTV core](https://forums.libretro.com/t/nvidia-shield-tv-simultaneous-button-presses-bug/20095/4) also.

**Comment by fr500** [comment]
It's a new menu but it doesn't change inputs at all

**Comment by ReploidX** [comment]
I uninstalled both RetroArch and RetroArch64; restarted; then installed RetroArch from Play.
RetroArch 1.7.5

- No settings changes other than downloading cores
- Using USB HORI Fighting Commander in PC and d-pad mode
- Also tried each "USB mode" on SATV, but none made a difference

Input Driver: android
Joypad Driver: android

I'm not sure what "retropad builtin core" referred to. For this controller only Modern and Classic were available, if that's what that meant. I don't have the SHIELD Controller. If Modern and Classic are known to not work I can try other controllers until one shows RetroPad.


**sfa3.zip**
FB Alpha 2012
FB Alpha 2012 CPS-2
MAME 2016
Throws rarely or never worked
Usually kick throws/lariats are more likely to work, but I feel this is something in SFA2/3. When I remapped punches to the kick buttons, punch throws/lariats still never reliably worked.


**Golden Axe (World).md**
Genesis Plus GX
Two-button special move rarely worked


**kof2000.zip**
FB Alpha 2012
Two-button sometimes worked - more than SF but still not reliably

**GameStream and Steam Link**
I tried SF30 through GameStream and Steam Link again, and was able to do throws and lariats consistently.


**SHIELD TV info:**
Model Number
P2897

Version
8.0.0

SW Version
7.2.2(30.7.130.7)

**Comment by ReploidX** [comment]
I've done some more testing on SHIELD TV with other controllers and cores, and for now the only thing I've found that helps is using Run-Ahead. I mentioned this on the Libretro forum and the problems with using it are still there, but with 1 frame of Run-Ahead I can at least do multi-button moves most of the time.

bsnes cores don't work well with Run-Ahead enabled (it goes down to roughly 40fps), but they also don't have the multi-button issue. Multi-buttons moves work in TMNT4 and SF without Run-Ahead enabled.

With Run-Ahead enabled, FBA and Genesis Plus GX can detect multi-button moves. FBA seems to have more visual glitches with Run-Ahead than Genesis Plus GX.

I also tested the same cores, ROMs, controllers on Windows and macOS and multi-button moves work without using Run-Ahead.

**Comment by brunobelo** [comment]
I'm having this problem, Shield TV, stock rom, no root, mess2014, mame2016 and mame cores, can't press 2 buttons at the same time. I found out playing Samurai Shodown for Neogeo CD, couldn't do any strong moves... I'm using an X360 wireless controller with wireless receiver.

**Comment by Immersion95** [comment]
Hi, I'm having the same problem on the Shield TV with FBA.

It works when you only press 2 buttons bu not when you do it with a command (shinku hadoken).

**Comment by orbea** [comment]
@fr500 Is the builtin retropad core even enabled on the shield? Anyone that can reproduce it should test it as it shows the buttons as they are pressed.

With linux its found at the very bottom of `Load Core`.

**Comment by ReploidX** [comment]
@orbea I tried the "Load Core" > "Start Remote RetroPad" with the HORI controller on macOS and SHIELD TV. It shows all the buttons working, but does it do anything more advanced than that?

Visually it looks like I can hit three buttons at exactly the same time. If it kept a record of what was pressed like in a fighting game training mode then I would know what it's reading "per frame."

I don't know what the right terminology would be, but with Run-Ahead turned on, is there more of an input buffer that's allowing it to be more lenient about how quickly it detects button presses?

**Comment by fr500** [comment]
hmmm ok I do think there's something going on.
While I can do throws in SFIII3, generally speaking I can't on SFA3.

The same moves are easy with the same pads on PC.

It's even worse for EX moves.

That said, the android input code hasn't changed so it may be a firmware issue.

**Comment by fr500** [comment]
Well I dunno about this really
whenever I feel like I'm getting some conclusions... I start pulling off ex hadoukens with great results.
As for the issue. Please stop with the runahead stuff... 

That said, runahead off, SFIII3n FBA, I can pull off throws 100% of the time with all controllers.
I only have problems with EX moves.

**Comment by brunobelo** [comment]
Runahead solved for FBA core, but not for mame or mess2014. I can perform the strong move on Samurai Shodown on NeoGeo (fba) but I can't do the same move on NeoGeo CD (mess2014 or mame).

**Comment by fr500** [comment]
mixing things never helps.
ie: changing all sort of unrelated settings when you're trying to solve it.

Anyway gonna make the polling block interval configurable to see if that changes the outcome.

**Comment by brunobelo** [comment]
I don't think that some specific symptom like that should be ignored, but that's me...

**Comment by fr500** [comment]
https://github.com/libretro/RetroArch/pull/8221

So once this is merged, look for input_block_timeout, and change it to... 8 for instance and see if it's any better.
If it works, then I'll add gui settings.

**Comment by fr500** [comment]
Try this apk please
https://bot.libretro.com/.radius/retroarch-release.apk

**Comment by fr500** [comment]
Good news, it works
Bad news it seems to give a performance hit

Good news, with the value at 3 it's pretty much spot on.

**Comment by fr500** [comment]
So this is fixed, but the fix causes performance issues.
Maybe let's keep it open? 

Or should I open a new one explaining what I did, and what does it cause?

**Comment by orbea** [comment]
It would probably be clearer in 6 months what the issue is with a new issue that linked this one.
